.github/workflows/scripts/generate-weekly-report.js

```diff
@@ -0,0 +1,428 @@
+      section.push(`<details>
+    <summary> Components </summary>\n`);
+      section.push(`${Object.keys(data).map((compName) => {
+        const {stability} = data[compName]
+        return `- [ ] ${compName}: ${JSON.stringify(stability)}`
+      }).join('\n')}
+</details>`);
+    }
+    section.push('</li>');
+    out.push(section.join('\n'));
+  }
+  out.push('</ul>');
+
+  // add json data
+  out.push('\n ## JSON Data');
+  out.push('<!-- MACHINE GENERATED: DO NOT EDIT -->');
+  out.push(`<details>
+<summary>Expand</summary>
+<pre>
+{
+  "issuesData": ${JSON.stringify(issuesData, null, 2)},
+  "componentData": ${JSON.stringify(componentData, null, 2)}
+}
+</pre>
+</details>`);
+  const report = out.join('\n');
+  return report;
+}
+
+async function createIssue({ github, lookbackData, report, context }) {
+  const title = `Weekly Report: ${lookbackData.sevenDaysAgo.toISOString().slice(0, 10)} - ${lookbackData.midnightYesterday.toISOString().slice(0, 10)}`;
+  return github.rest.issues.create({
+    owner: context.payload.repository.owner.login,
+    repo: "opentelemetry-collector-contrib",
+    title,
+    body: report,
+    labels: ["report"]
+  })
+}
+
+async function getLastWeeksReport({ github, since, context }) {
+  const issues = await github.rest.issues.listForRepo({
+    owner: context.payload.repository.owner.login,
+    repo: 'opentelemetry-collector-contrib',
+    state: 'all', // To get both open and closed issues
+    labels: ["report"],
+    since: since.toISOString(),
+    per_page: 1,
+    direction: "asc"
+  });
+  if (issues.data.length === 0) {
+    return null;
+  } 
+  // grab earliest issue if multiple
+  return issues.data[0];
+}
+
+function parseJsonFromText(text) {
+  // Use regex to find the JSON data within the <pre></pre> tags
+  const regex = /<pre>\s*([\s\S]*?)\s*<\/pre>/;
+  const match = text.match(regex);
+
+  if (match && match[1]) {
+    // Parse the found string to JSON
+    return JSON.parse(match[1]);
+  } else {
+    throw new Error("JSON data not found");
+  }
+}
+
+async function processIssues({ github, context, lookbackData }) {
+  const issuesData = await getIssuesData({github, context});
+
+  const prevReportLookback = new Date(lookbackData.sevenDaysAgo)
+  prevReportLookback.setDate(prevReportLookback.getDate() - 7)
+  const previousReportIssue = await getLastWeeksReport({github, since: prevReportLookback, context});
+  // initialize to zeros
+  let previousReport = null;
+
+  if (previousReportIssue !== null) {
+    const {created_at, id, url, title} = previousReportIssue;
+    debug({ msg: "previous issue", created_at, id, url, title })
+    previousReport = parseJsonFromText(previousReportIssue.body)
+  }
+
+  return {issuesData, previousReport}
+
+
+}
+
+const findFilesByName = (startPath, filter) => {
+  let results = [];
+
+  // Check if directory exists
+  let files;
+  try {
+      files = fs.readdirSync(startPath);
```

Comment:
Why not use the [async](https://nodejs.org/docs/latest-v18.x/api/fs.html#fspromisesreaddirpath-options) versions of the fs functions?

Reply:
the logic was easier to follow when sticking with sync. since the number of directories isn't that large, felt like this wouldn't be a bottleneck

Reply:
> the logic was easier to follow when sticking with sync.

Not sure I understand. The flow of the logic is the same with async, you just add the `await` keyword.

```javascript
    files = await fs.readdir(startPath);
```

The `await` keyword is already used elsewhere in the script, e.g.

```javascript
const response = await octokit.issues.listForRepo(queryParams);
```

I thought maybe it's an issue in importing the `fs/promises` version of the `fs` library?